src/mse.cu

```diff
@@ -0,0 +1,602 @@
+            break;
+
+        case 32:
+            if (incx == 1 && incy == 1) {
+                mse_error_kernel1<T, 32, Reduce><<<numBlocks, numThreads, sharedSize>>>(n, output, labels, r_output);
+            } else {
+                mse_error_kernel<T, 32, Reduce><<<numBlocks, numThreads, sharedSize>>>(n, output, incx, labels, incy, r_output);
+            }
+            break;
+
+        case 16:
+            if (incx == 1 && incy == 1) {
+                mse_error_kernel1<T, 16, Reduce><<<numBlocks, numThreads, sharedSize>>>(n, output, labels, r_output);
+            } else {
+                mse_error_kernel<T, 16, Reduce><<<numBlocks, numThreads, sharedSize>>>(n, output, incx, labels, incy, r_output);
+            }
+            break;
+
+        case 8:
+            if (incx == 1 && incy == 1) {
+                mse_error_kernel1<T, 8, Reduce><<<numBlocks, numThreads, sharedSize>>>(n, output, labels, r_output);
+            } else {
+                mse_error_kernel<T, 8, Reduce><<<numBlocks, numThreads, sharedSize>>>(n, output, incx, labels, incy, r_output);
+            }
+            break;
+
+        case 4:
+            if (incx == 1 && incy == 1) {
+                mse_error_kernel1<T, 4, Reduce><<<numBlocks, numThreads, sharedSize>>>(n, output, labels, r_output);
+            } else {
+                mse_error_kernel<T, 4, Reduce><<<numBlocks, numThreads, sharedSize>>>(n, output, incx, labels, incy, r_output);
+            }
+            break;
+
+        case 2:
+            if (incx == 1 && incy == 1) {
+                mse_error_kernel1<T, 2, Reduce><<<numBlocks, numThreads, sharedSize>>>(n, output, labels, r_output);
+            } else {
+                mse_error_kernel<T, 2, Reduce><<<numBlocks, numThreads, sharedSize>>>(n, output, incx, labels, incy, r_output);
+            }
+            break;
+
+        case 1:
+            if (incx == 1 && incy == 1) {
+                mse_error_kernel1<T, 1, Reduce><<<numBlocks, numThreads, sharedSize>>>(n, output, labels, r_output);
+            } else {
+                mse_error_kernel<T, 1, Reduce><<<numBlocks, numThreads, sharedSize>>>(n, output, incx, labels, incy, r_output);
+            }
+            break;
+    }
+}
+
+template <bool Loss, typename T>
+T mse_kernel_run(size_t n, const T* output, size_t incx, const T* labels, size_t incy) {
+    T result = 0;
+
+    const size_t cpu_threshold = Loss ? 1 : 1024;
+
+    if (Loss && n < cpu_threshold && incx == 1 && incy == 1) {
+        T* host_output = new T[n];
+        T* host_labels = new T[n];
+
+        cuda_check(cudaMemcpy(host_output, output, n * sizeof(T), cudaMemcpyDeviceToHost));
+        cuda_check(cudaMemcpy(host_labels, labels, n * sizeof(T), cudaMemcpyDeviceToHost));
+
+        for (size_t i = 0; i < n; i++) {
+            result += (host_labels[i] - host_output[i]) * (host_labels[i] - host_output[i]);
+        }
+        result /= n;
+
+        delete[] host_output;
+        delete[] host_labels;
```

Comment:
There should be a `return result` here. If you want, you can also fix it in BCE where it's missing.